packages/modelviewer.dev/examples/postprocessing/index.html

```diff
@@ -0,0 +1,394 @@
+            effects is to use the default <code>selection</code> provided by the <code>&lt;effect-composer&gt;</code>, which contains a set of
+            all meshes in the model.<br/>
+            By providing an array of mesh names to the <code>selection</code> attribute, you may enable the effects only for those meshes.<br/>
+            This example demonstrates the application of the <code>&lt;selective-bloom-effect&gt;</code> exclusively to Meshes with an emissive material:</p>
+          </div>
+          <example-snippet stamp-to="selective-effects" highlight-as="html">
+            <template>
+<model-viewer camera-orbit="45deg 115deg auto" camera-controls touch-action="pan-y" autoplay src="../../shared-assets/models/RocketShip.glb" ar alt="A 3D model of a Rocket Ship">
+  <effect-composer id="selectiveComposer" render-mode="quality">
+    <selective-bloom-effect strength="5" radius="0.5" threshold="0"></selective-bloom-effect>
+  </effect-composer>
+</model-viewer>
+<script type="module">
+  const selectiveComposer = document.querySelector("effect-composer#selectiveComposer");
+  const selectiveBloom = selectiveComposer.querySelector("selective-bloom-effect");
+
+  selectiveComposer.addEventListener('updated-selection', () => {
+    const selection = [];
+    selectiveComposer.selection.forEach((obj) => 
+      (obj.material.emissive.r > 0 
+      || obj.material.emissive.g > 0 
+      || obj.material.emissive.b > 0) &&
+        selection.push(obj)
+    );
+    selectiveBloom.selection = selection;
+  });
+</script>
+            </template>
+          </example-snippet>
+        </div>
+      </div>
+    </div>
+
+    <div class="sample">
+      <div id="combining-effects" class="demo"></div>
+      <div class="content">
+        <div class="wrapper">
+          <div class="heading">
+            <h2 class="demo-title">Combining Effects</h2>
+            <p>&lt;model-viewer-effects&gt;</code> will attempt to join as many of your effects as possible into a single
+            <code>EffectPass</code>, which has a significant performance benefit as opposed to keeping all the effects
+            in separates Passes. Effects can be disabled by setting the <code>blend-mode</code> to <code>skip</code>.
+            <br>
+            It is recommended to follow the official <a href="https://github.com/pmndrs/postprocessing/wiki/Effect-Merging#effect-execution-order" target='_blank'>Postprocessing Wiki</a>
+            for best practices in combining effects. The <code>&lt;model-viewer-effects&gt;</code> handles effect compatibility issues by 
+            placing incompatible effects in a separate Pass.
+            </p>
+          </div>
+          <example-snippet stamp-to="combining-effects" highlight-as="html">
+            <template>
+<model-viewer id="effectsViewer" camera-controls touch-action="pan-y" src="../../shared-assets/models/EmissiveStrengthTest.glb" ar alt="A 3D ">
```

Comment:
Even when an empty effect composer is used, the rendering is not actually using the Three render function, rather the base RenderPass of the effect composer. As I mentioned in some earlier comment, the default antialiasing in your renderer isn't really active then.

Reply:
I think the effect on the depth buffer is the critical thing to understand. Somehow firefox and windows are unaffected, but chrome and safari (on mac) are. I'd also like to see examples of the ugly artifacting with certain effects. I'm curious if it's related, and also if it varies by platform.

Reply:
![image](https://user-images.githubusercontent.com/54665621/230218210-13ce8c62-ed61-45e4-aead-392894abca44.png)

Artifacts like this for example in selective-bloom, which uses the depth buffer to choose which objects are emissive. Seems that the msaa solution provided by the EffectComposer causes artifacts in the depthBuffer, which confuses the effect and makes the pixels glow. 

This happens also if the alternative smaa-effect is applied before the depth based effect, for example here is the same example with msaa disabled and with smaa placed before the selective bloom:
![image](https://user-images.githubusercontent.com/54665621/230218979-3d9b9038-dced-45c2-8fc1-1f1d617657ae.png)

And here again with msaa disabled, and smaa placed *after* the selective-bloom:
![image](https://user-images.githubusercontent.com/54665621/230219029-2d232e91-ff9f-4ede-afc4-56a1782179a6.png)

Finally, both msaa and smaa disabled for reference:
![image](https://user-images.githubusercontent.com/54665621/230219152-815c40fa-5a21-402f-81e2-9a108553c8fd.png)

Reply:
These artifacts occur because the masking technique of the `SelectiveBloomEffect` relies on depth which contains aliasing artifacts. When the color image is antialiased, the depth mask doesn't match the pixels anymore which causes these artifacts to linger. The solution is to render the selected objects to a separate render target during the main geometry render step via MRT instead of masking after the fact.

MSAA causes more fundamental issues for depth-based image effects like SSAO or DoF and, as far as I know, the solution is to implement a custom MSAA resolve step, but that's not possible in WebGL. This [wiki entry](https://github.com/pmndrs/postprocessing/wiki/Antialiasing) shows the artifacts that occur when rendering SSAO with MSAA enabled. This [article](https://mtnphil.wordpress.com/2013/06/26/know-your-ssao-artifacts/) touches on the subject as well (further down, heading "MSAA and SSAO together").

Reply:
@vanruesc Thanks for the info, and it seems we get pretty nice results with the selections @Beilinson has posted for our examples. However, do you have any insight regarding the fairly extreme depth buffer imprecision artifacts we're getting above (the astronaut)? It seems very odd that it's browser-dependent.

Reply:
@vanruesc Could you maybe explain why there seems to be less issues with SMAA artifacting when the outputEncoding is set to linearEncoding (both examples are without msaa, although that causes more or less the same issues obviously). setting the frameBufferType to UnsignedByteType or HalfFloatType doesn't make a difference.

Case in point (sRGBEncoding):
![image](https://user-images.githubusercontent.com/54665621/230233822-c0b7000f-0ef9-4312-8ac7-17a59bd190af.png)

vs linearEncoding:
![image](https://user-images.githubusercontent.com/54665621/230233675-0cb575ee-5be1-4b60-8d31-0ded204d8344.png)

Reply:
>  So the question is, what is different about our rendering when we have an empty effect composer, compared to no effect-composer?

I'm not sure, but you might be rendering directly to the canvas with an "empty" effect composer (see [autoRenderToScreen](https://pmndrs.github.io/postprocessing/public/docs/class/src/core/EffectComposer.js~EffectComposer.html#instance-member-autoRenderToScreen)) in which case `antialias` might be `false`.

> However, do you have any insight regarding the fairly extreme depth buffer imprecision artifacts we're getting above (the astronaut)? It seems very odd that it's browser-dependent.

This could be a driver issues as you said. Or it could be related to depth precision. Postprocessing uses `UnsignedIntType` for depth textures by default which can technically result in 16 bit precision although it really shouldn't. Try enabling the stencil buffer: `new EffectComposer(renderer, { stencilBuffer: true })` to force 24 bit precision via `UnsignedInt248Type`.

> Could you maybe explain why there seems to be less issues with SMAA artifacting when the outputEncoding is set to linearEncoding

The aliasing artifacts are probably very similar in shape, but less noticable with darker colors. Are you using a transparent canvas? Postprocessing effects struggle with alpha because WebGL doesn't provide a way to properly blend colors with the background.

Reply:
> I'm not sure, but you might be rendering directly to the canvas with an "empty" effect composer (see [autoRenderToScreen](https://pmndrs.github.io/postprocessing/public/docs/class/src/core/EffectComposer.js~EffectComposer.html#instance-member-autoRenderToScreen)) in which case `antialias` might be `false`.

Actually that fixed it, the issue was that an "empty" effect composer actually had more than just a renderPass, now with only a renderPass it is rendering identically to no effectComposer at all. Thanks!

> This could be a driver issues as you said. Or it could be related to depth precision. Postprocessing uses `UnsignedIntType` for depth textures by default which can technically result in 16 bit precision although it really shouldn't. Try enabling the stencil buffer: `new EffectComposer(renderer, { stencilBuffer: true })` to force 24 bit precision via `UnsignedInt248Type`.

I have pushed a version with `stencilBuffer: true`, on my machine despite not getting such extreme artifacts like @elalish posted I am getting minor white pixel artifacts like in the outputEncoding example above. Not sure if it'll fix the more extreme mac artifacting

> The aliasing artifacts are probably very similar in shape, but less noticable with darker colors. Are you using a transparent canvas? Postprocessing effects struggle with alpha because WebGL doesn't provide a way to properly blend colors with the background.
 
Yes it is using a transparent canvas, I also noticed that setting `alpha: false` fixes the artifacts but thats not an option for release because there is one WebGLRenderer instance for all the scenes. 

I've also attempted https://discourse.threejs.org/t/using-post-processing-through-effectcomposer-causes-jagged-artifacts-on-model-browser-safari/22918/5 as this and one or two other discussions suggested setting the `logarithmicDepthBuffer: true`, again I'm not sure if it'll fix the mac issues, but it doesn't fix the white pixel outline artifacts from my examples.

I'm mostly okay with the white artifacts in my example, they are really only noticeable on a light background with the outline-effect. The Mac artifacting is a much bigger problem however, hopefully the `stencilBuffer` fixes it